lib/helper.js

```diff
@@ -3,15 +3,132 @@
+            }
+            return found || []
+          }
+        }
+        // lots of checks look at the details of the first entry, which is
+        // usually 'Mozilla', but not always (e.g. 'Opera'),
+        // so record the first entry separately:
+        if (!result.firstEntry) {
+          result.firstEntry = result[key]
+        }
+      }
+    }
+  }
+  if (!result.firstEntry) {
+    // ensure firstEntry is always set no matter what (null object pattern)
+    result.firstEntry = {
+      version: null,
+      details: [],
+      hasDetail: () => false,
+      getDetail: () => []
+    }
+  }
+  return result
+}
+
+const extractFromParts = (parts, checks) => checks
+  .map((check) => check(parts))
+  .filter((result) => result)[0] || []
+
+const extractMacVersion = (m) => m[1].replace(/_/g, '.')
+
+const WINDOWS_NT_VERSION_MAP = {
+  5.1: 'XP',
+  5.2: 'XP',
+  '6.0': 'Vista',
+  6.1: '7',
+  6.2: '8',
+  6.3: '8.1',
+  6.4: '10',
+  '10.0': '10'
+}
+const extractWindowsVersion = (m) => WINDOWS_NT_VERSION_MAP[m[1]]
+
+const UA_BROWSERS = [
+  ({ phantomjs }) => phantomjs && // also contains Safari
+    ['PhantomJS', phantomjs.version],
+
+  ({ headlesschrome }) => headlesschrome && // also contains Safari
+    ['Chrome Headless', headlesschrome.version],
+
+  ({ opera, version }) => opera &&
+    ['Opera', version && version.version],
+
+  ({ firefox }) => firefox &&
+    ['Firefox', firefox.version],
+
+  ({ edg }) => edg && // also contains Chrome, Safari
+    ['Edge', edg.version],
+
+  ({ chrome }) => chrome && // also contains Safari
+    ['Chrome', chrome.version],
+
+  ({ firstEntry, version }) => firstEntry.hasDetail(/^iphone/i) && // also contains Safari
+    ['Mobile Safari', version && version.version],
+
+  ({ firstEntry, version }) => firstEntry.hasDetail(/^android/i) && // also contains Safari
+    ['Android Browser', version && version.version],
+
+  ({ safari, version }) => safari &&
+    ['Safari', version && version.version],
+
+  ({ firstEntry }) => firstEntry.hasDetail(/^msie/i) &&
+    ['IE', firstEntry.getDetail(/^msie ([\d.]+)/i)[1]]
+]
+
+const UA_SYSTEMS = [
+  ({ firstEntry }) => firstEntry.hasDetail(/^android/i) &&
+    ['Android', firstEntry.getDetail(/^android ([\d.]+)/i)[1]],
+
+  ({ firstEntry }) => firstEntry.hasDetail(/^iphone/i) &&
+    ['iOS', firstEntry.getDetail(/iphone os ([\d._]+)/i, extractMacVersion)],
+
+  ({ ubuntu }) => ubuntu &&
+    ['Ubuntu', ubuntu.version],
+
+  ({ firstEntry }) => firstEntry.hasDetail(/^freebsd/i) &&
+    ['FreeBSD', null],
+
+  ({ firstEntry }) => firstEntry.hasDetail(/^linux/i) &&
+    ['Linux', firstEntry.getDetail(/^linux (.+)/i)[1]],
+
+  ({ firstEntry }) => firstEntry.hasDetail(/mac os/i) &&
+    ['Mac OS', firstEntry.getDetail(/mac os(?: x)? ([\d._]+)/i, extractMacVersion)],
+
+  ({ firstEntry }) => firstEntry.hasDetail(/^windows/i) &&
+    ['Windows', firstEntry.getDetail(/windows nt ([\d.]+)/i, extractWindowsVersion)]
+]
+
```

Comment:
Where did you get all this info? Probably from the implementation of `ua-parser-js`? Can we add a reference?

Reply:
Also, can you please add comments to the functions?

Reply:
source of this is actually the test user-agent strings which were already in this project. The code and regular expressions here have no relation to ua-parser-js's code.

Reply:
As for documentation: if the team is happy with this approach, I can extract the code into its own file, add finer-grained tests, and document. Wanted to get a quick first pass submitted without all that so that I didn't waste too much time if the team was opposed to the whole idea.

Reply:
I am happy with your approach. It is very elegant! :)
I'd like to hear from @devoto13 as well.

Reply:
Approach overall LGTM!
I'll try to look more into details in the coming days.